dokka-integration-tests/gradle/src/main/kotlin/org/jetbrains/dokka/it/gradle/junit/TestedVersions.kt

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package org.jetbrains.dokka.it.gradle.junit
+
+import org.jetbrains.dokka.it.gradle.utils.SemVer
+import org.jetbrains.dokka.it.optionalSystemProperty
+import org.jetbrains.dokka.it.systemProperty
+
+
+/**
+ * The parameters used to run each invocation of a [DokkaGradlePluginTest] test.
+ *
+ * [DokkaGradlePluginTestExtension] generates all possible values, and creates a test invocation
+ * for each valid combination.
+ *
+ * To access the specific values used for a test invocation, add a [TestedVersions]
+ * parameter in the test function.
+ */
+data class TestedVersions(
+    /** Kotlin Gradle Plugin version. */
+    val kgp: SemVer,
+    /** Gradle version. */
+    val gradle: SemVer,
+    /** Dokka Gradle Plugin version. */
+    val dgp: SemVer,
+    /**
+     * The Android Gradle Plugin version being tested.
+     *
+     * The test must be annotated with [TestsAndroidGradlePlugin] for this value to be set.
+     */
+    val agp: SemVer?,
+    /**
+     * The Compose version being tested.
+     *
+     * The test must be annotated with [TestsCompose] for this value to be set.
+     */
+    val composeGradlePlugin: SemVer?,
+) : Comparable<TestedVersions> {
+
+    constructor(
+        kgp: String,
+        gradle: String,
+        dgp: String,
+        agp: String?,
+        composeGradlePlugin: String?,
+    ) : this(
+        kgp = SemVer(kgp),
+        gradle = SemVer(gradle),
+        dgp = SemVer(dgp),
+        agp = agp?.let { SemVer(it) },
+        composeGradlePlugin = composeGradlePlugin?.let { SemVer(it) },
+    )
+
+    /**
+     * Check if the combination of versions is valid.
+     *
+     * For example, AGP 7 must use Gradle 7. If [agp] is `8.0.0` and [gradle] is `7.0.0`
+     * then the combination is invalid.
+     */
+    private fun isValid(): Boolean {
+        if (agp != null) {
+            if (agp.major != gradle.major) {
+                return false
+            }
+
+            when (kgp) {
+                SemVer("1.9.0") -> {
+                    if (agp !in SemVer("7.0.0")..SemVer("8.2.0")) {
+                        return false
+                    }
+                }
+            }
+        }
+
+        return true
+    }
+
+    private val versionsMap: Map<String, SemVer> =
+        buildMap {
+            put("Dokka", dgp)
+            put("Kotlin", kgp)
+            put("Gradle", gradle)
+            if (agp != null) put("Android", agp)
+            if (composeGradlePlugin != null) put("Compose", composeGradlePlugin)
+        }
+
+    /**
+     * Join the major and minor versions of each version.
+     * Used by [allTestedVersions] to filter out duplicated patch versions.
```

Comment:
it’s not clear for me, why this could be needed?
If we do have different `patch` versions, may be we do need to test them? E.g may be `2.1.21` introduces bug-fix which is required for Dokka

Reply:
The main purpose is to prevent unnecessary testing. We don't need to test both KGP 1.9.23 *and* KGP 1.9.24.

However, we do need to consider patch numbers when generating test versions, because Compose has very strict requirements for specific Kotlin versions, down to the patch level. So Compose compiler plugin 1.5.14 is only compatible with KGP 1.9.24, and Compose compiler 1.5.13 is only compatible with KGP 1.9.23.

https://developer.android.com/jetpack/androidx/releases/compose-kotlin


So, in the list of potential KGP versions we have to have the specific versions required by Compose compiler. Otherwise it might not be able to generate any valid combination for Compose and KGP.

When Compose is tested, then it will restrict the KGP version so there's only one KGP version.

However, when Compose isn't being tested, there's no restriction. The test infra will generate a list of all versions, which will contain combinations for both KGP 1.9.23 *and* KGP 1.9.24.

But for Dokka we don't really care about testing KGP 1.9.23 *and* KGP 1.9.24. So we filter out the lower patch versions.

Does that make sense? It's a bit hard to explain. If it helps, try removing the filtering and see that there's a lot of extra testing.

> may be 2.1.21 introduces bug-fix which is required for Dokka

If at some point DGP requires a specific version of KGP, then let's think of a solution then. We could add a custom requirement to make sure DGP gets tested.

Reply:
I've added a provider, `TestedVersionsSource`, that uses a Sequence builder to generate the versions.

This can be used to explicitly add a new combination of versions, if necessary. But for now, it just needs to provide the combinations of all tested versions.

Does this work better?